src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/wsdl/WsdlParser.java

```diff
@@ -282,22 +288,35 @@ public ServiceInfo(String name, String title, String url,
             this.wsdlUrl = wsdlUrl;
         }
 
+        // cannot change the method signature when overriding so we need to sneakily throw the WsdlParseException
+        @SneakyThrows(WsdlParseException.class)
         @Override
         public InputSource getBaseInputSource() {
-            try {
+            try (ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream()) {
                 URLConnection conn = new URL(wsdlUrl).openConnection();
                 if (conn instanceof HttpsURLConnection) {
                     configureHttps((HttpsURLConnection) conn);
                 }
 
-                // cache the response
-                byte[] response;
                 try (InputStream in = conn.getInputStream()) {
-                    response = IOUtils.toByteArray(in);
+                    long count = 0;
+                    int n;
+                    byte[] buf = new byte[BUF_SIZE];
+                    while ((n = in.read(buf)) != -1) {
+                        count += n;
+                        if (count > MAX_DESCRIPTION_SIZE) {
+                            throw new WsdlParseException(
+                                    "Error reading WSDL: Size exceeds " + MAX_DESCRIPTION_SIZE + " bytes.");
+                        }
+                        byteArrayOutputStream.write(buf, 0, n);
+                    }
                 }
+                byte[] response = byteArrayOutputStream.toByteArray();
                 log.trace("Received WSDL response: {}", new String(response));
 
                 return new InputSource(new ByteArrayInputStream(response));
+            } catch (WsdlParseException e) {
+                throw e;
             } catch (Throwable t) {
```

Comment:
This catches also java.lang.Errors, which should not be normally done. Should be narrowed to the actual possible checked exception types or at least "Exception".

Reply:
Updated to catch `Exception`s